lib/util/id3_utils.js

```diff
@@ -0,0 +1,298 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+goog.provide('shaka.util.Id3Utils');
+
+goog.require('shaka.log');
+goog.require('shaka.util.BufferUtils');
+goog.require('shaka.util.StringUtils');
+
+
+/**
+ * @summary A set of Id3Utils utility functions.
+ * @export
+ */
+shaka.util.Id3Utils = class {
+  /**
+   * @param {Uint8Array} data
+   * @param {number} offset
+   * @return {boolean}
+   * @private
+   */
+  static isHeader_(data, offset) {
+    /*
+     * http://id3.org/id3v2.3.0
+     * [0]     = 'I'
+     * [1]     = 'D'
+     * [2]     = '3'
+     * [3,4]   = {Version}
+     * [5]     = {Flags}
+     * [6-9]   = {ID3 Size}
+     *
+     * An ID3v2 tag can be detected with the following pattern:
+     *  $49 44 33 yy yy xx zz zz zz zz
+     * Where yy is less than $FF, xx is the 'flags' byte and zz is less than $80
+     */
+    if (offset + 10 <= data.length) {
+      // look for 'ID3' identifier
+      if (data[offset] === 0x49 &&
+          data[offset + 1] === 0x44 &&
+          data[offset + 2] === 0x33) {
+        // check version is within range
+        if (data[offset + 3] < 0xff && data[offset + 4] < 0xff) {
+          // check size is within range
+          if (data[offset + 6] < 0x80 &&
+              data[offset + 7] < 0x80 &&
+              data[offset + 8] < 0x80 &&
+              data[offset + 9] < 0x80) {
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * @param {Uint8Array} data
+   * @param {number} offset
+   * @return {boolean}
+   * @private
+   */
+  static isFooter_(data, offset) {
+    /*
+     * The footer is a copy of the header, but with a different identifier
+     */
+    if (offset + 10 <= data.length) {
+      // look for '3DI' identifier
+      if (data[offset] === 0x33 &&
+          data[offset + 1] === 0x44 &&
+          data[offset + 2] === 0x49) {
+        // check version is within range
+        if (data[offset + 3] < 0xff && data[offset + 4] < 0xff) {
+          // check size is within range
+          if (data[offset + 6] < 0x80 &&
+              data[offset + 7] < 0x80 &&
+              data[offset + 8] < 0x80 &&
+              data[offset + 9] < 0x80) {
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * @param {Uint8Array} data
+   * @param {number} offset
+   * @return {number}
+   * @private
+   */
+  static readSize_(data, offset) {
+    let size = 0;
+    size = (data[offset] & 0x7f) << 21;
+    size |= (data[offset + 1] & 0x7f) << 14;
+    size |= (data[offset + 2] & 0x7f) << 7;
+    size |= data[offset + 3] & 0x7f;
+    return size;
+  }
+
+  /**
+   * @param {Uint8Array} data
+   * @return {shaka.extern.MetadataRawFrame}
+   * @private
+   */
+  static getFrameData_(data) {
```

Comment:
If "unsynchronisation" (https://id3.org/id3v2.3.0#sec5) or the "extended header" (https://id3.org/id3v2.3.0#ID3v2_extended_header) are not expected in the input, you should document somewhere why, and that they are ignored.

If either of those features might show up in the data, you should probably check the flags and throw or something when they are present.  Otherwise, the data will be at offsets other than what you expect, and this parser will be broken.

Reply:
I will add a comment, because as far as I have seen in HLS with TS and ACC this does not happen. Since ID3 works in many formats, but in our case it is not necessary.